### PR TITLE
29: implement gas refund for voting

### LINKED ISSUE #29: Gas refunds
Implement gas refund for voting from community treasury.

Approach:

1. enable receive callback on CollectiveGovernance
2. disable fallback on CollectiveGovernance
3. in voting methods, if contract balance > gas and less than MAXIMUM_REFUND, transfer that amount from contract to voter address
4. to enable gas refund community should transfer ETH to governance contract

Suggestions:
1. require sender == origin for refund - voter is a wallet and not a contract

### PR DESCRIPTION
**Pull request by jac18281828**
closes #29 
* implements the gas refund when contract is capitalized 
* storage size and cleanup changes for code size limit

### CODE REVIEW

File: contracts/CollectiveGovernance.sol [hunk 1]
```diff
@@ -552,13 +635,68 @@ contract CollectiveGovernance is Governance, VoteStrategy, ERC165 {
...
+        }
+    }
+
+    function sendGasRebate(address recipient, uint256 startGas) internal {
+        uint256 balance = getRebateBalance();
+        if (balance == 0) {
+            return;
+        }
+
+        uint256 totalGasUsed = startGas - gasleft();
```
Comment (dannyadkins) [inline contracts/CollectiveGovernance.sol:691]: should we move this down two lines to maximize the gas refunded to user? otherwise they have to pay for those two lines

File: contracts/CollectiveGovernance.sol [hunk 2]
```diff
@@ -552,13 +635,68 @@ contract CollectiveGovernance is Governance, VoteStrategy, ERC165 {
...
+    function _undoVote(uint256 _proposalId, uint256 _tokenId) private {
+        uint256 count = _storage.undoVoteById(_proposalId, msg.sender, _tokenId);
+        if (count > 0) {
+            emit VoteUndo(_proposalId, msg.sender, count);
+        } else {
+            revert("Not voter");
+        }
+    }
+
+    function sendGasRebate(address recipient, uint256 startGas) internal {
```
Comment (dannyadkins) [inline contracts/CollectiveGovernance.sol:676]: should we check here that gas rebates are actually enabled? or if the contract has 0 ETH balance, I guess it won't be able to transfer anything anyway. But i could imagine a setup where someone is using the contract as a community treasury (even though we are going to provide a separate contract) without necessarily wanting to refund voters... since _maximumBaseFeeRefund and _maximumGasUsedRefund are > 0 by default, calling this would still trigger a refund right? 

so, in short, maybe we have a param gasRefundEnabled?

Comment (dannyadkins) [inline contracts/CollectiveGovernance.sol:676]: what's the reason for this being internal and not private?

File: contracts/CollectiveGovernance.sol (hunk 1 above)
Comment (jac18281828) [reply contracts/CollectiveGovernance.sol:691]: I could go either way but it seems like a trivial maybe even not measurable change to the gas used, while on the other hand seemingly lowering the readability and understandability of this code.    

> should we move this down two lines to maximize the gas refunded to user? otherwise they have to pay for those two lines

File: contracts/CollectiveGovernance.sol (hunk 2 above)
Comment (jac18281828) [reply contracts/CollectiveGovernance.sol:676]: Of course it can be private.   When I thought about it for a few seconds I considered that any future contract that extends this one might also want to send refunds. 

> what's the reason for this being internal and not private?

Comment (jac18281828) [reply contracts/CollectiveGovernance.sol:676]: Yes 0 eth balance is disabled.   There are plenty of other options for a community treasury including a vault and the time lock.    This usage is pretty consistent with how this is implemented by others as well, not that we have to follow that but it seems a reasonable approach to me.   

I think it makes sense to have the treasury separated off given the importance of that function and the idea that a hack to the vote may not ultimately be exploitable.

**Review by dannyadkins — APPROVED** [review]

**Review by jamalh110 — APPROVED** [review]

File: contracts/CollectiveGovernance.sol (hunk 1 above)
Comment (jac18281828) [reply contracts/CollectiveGovernance.sol:691]: I moved it down leaving some whitespace.  Hopefully it's readable enough.

File: contracts/GovernanceBuilder.sol [hunk 3]
```diff
@@ -167,34 +170,50 @@ contract GovernanceBuilder is GovernanceCreator, ERC165 {
...
+        transferOwnership(_timeLock, _governanceAddress);
         transferOwnership(_storage, _governanceAddress);
+        _governanceContractRegistered[_governanceAddress] = true;
         emit GovernanceContractCreated(_creator, _properties.name, address(_storage), _governanceAddress);
         return _governanceAddress;
     }
 
+    /// @notice identify a contract that was created by this builder
+    /// @return bool True if contract was created by this builder
+    function contractRegistered(address _contract) external view returns (bool) {
```
Comment (jac18281828) [inline contracts/GovernanceBuilder.sol:213]: @dannyadkins contract registration
